File: crates/ocr-engine/src/coercion.rs
use crate::schema::{DataType, ResponseSchema};
use serde_json::{json, Map, Value};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CoercionError {
    #[error("Missing required field: {0}")]
    MissingRequiredField(String),
    #[error("Type mismatch for field {0}: expected {1}")]
    TypeMismatch(String, String),
    #[error("Invalid enum value '{0}' for field {1}. Allowed values: {2}")]
    InvalidEnumValue(String, String, String),
}

/// Validates raw JSON returned by AI Vision Model tool call against ResponseSchema and coerces data types
pub fn validate_and_coerce(schema: &ResponseSchema, raw_value: Value) -> Result<Value, CoercionError> {
    let raw_map = raw_value.as_object().ok_or_else(|| {
        CoercionError::TypeMismatch("root".to_string(), "JSON object".to_string())
    })?;

    let mut result_map = Map::new();

    for (field_name, field_def) in &schema.properties {
        let value_opt = raw_map.get(field_name);

        match (value_opt, field_def.required) {
            (None, true) => return Err(CoercionError::MissingRequiredField(field_name.clone())),
            (None, false) => {
                if let Some(def_val) = &field_def.default {
                    result_map.insert(field_name.clone(), def_val.clone());
                } else {
                    result_map.insert(field_name.clone(), Value::Null);
                }
            }
            (Some(val), _) => {
                let coerced_val = coerce_field_value(field_name, &field_def.r#type, val)?;
                result_map.insert(field_name.clone(), coerced_val);
            }        }
    }

    Ok(Value::Object(result_map))
}

fn coerce_field_value(field_name: &str, data_type: &DataType, val: &Value) -> Result<Value, CoercionError> {
    match data_type {
        DataType::String | DataType::Text | DataType::Email | DataType::Phone | DataType::Date | DataType::DateTime | DataType::Currency => {
            match val {
                Value::String(s) => Ok(json!(s.trim())),
                Value::Number(n) => Ok(json!(n.to_string())),
                Value::Bool(b) => Ok(json!(b.to_string())),
                Value::Null => Ok(Value::Null),
                _ => Err(CoercionError::TypeMismatch(field_name.to_string(), "string".to_string())),
            }
        }
        DataType::Integer => match val {
            Value::Number(n) if n.is_i64() => Ok(json!(n.as_i64().unwrap())),
            Value::Number(n) if n.is_f64() => Ok(json!(n.as_f64().unwrap() as i64)),
            Value::String(s) => s.trim().parse::<i64>().map(|i| json!(i)).map_err(|_| {
                CoercionError::TypeMismatch(field_name.to_string(), "integer".to_string())
            }),
            _ => Err(CoercionError::TypeMismatch(field_name.to_string(), "integer".to_string())),
        },
        DataType::Number => match val {
            Value::Number(n) => Ok(json!(n.as_f64().unwrap_or(0.0))),
            Value::String(s) => s.trim().parse::<f64>().map(|f| json!(f)).map_err(|_| {
                CoercionError::TypeMismatch(field_name.to_string(), "number".to_string())
            }),
            _ => Err(CoercionError::TypeMismatch(field_name.to_string(), "number".to_string())),
        },
        DataType::Boolean => match val {
            Value::Bool(b) => Ok(json!(b)),
            Value::String(s) => match s.trim().to_lowercase().as_str() {
                "true" | "1" | "yes" => Ok(json!(true)),
                "false" | "0" | "no" => Ok(json!(false)),
                _ => Err(CoercionError::TypeMismatch(field_name.to_string(), "boolean".to_string())),
            },
            _ => Err(CoercionError::TypeMismatch(field_name.to_string(), "boolean".to_string())),
        },
        DataType::Enum { values } => {
            let str_val = match val {
                Value::String(s) => s.trim(),
                _ => return Err(CoercionError::TypeMismatch(field_name.to_string(), "enum string".to_string())),
            };
            for allowed in values {
                if allowed.eq_ignore_ascii_case(str_val) {
                    return Ok(json!(allowed));
                }
            }
            Err(CoercionError::InvalidEnumValue(
                str_val.to_string(),
                field_name.to_string(),
                values.join(", "),
            ))
        }
        DataType::Object { properties } => {
            let sub_schema = ResponseSchema::new(properties.clone());
            validate_and_coerce(&sub_schema, val.clone())
        }
        DataType::Array { items } => {
            let arr = val.as_array().ok_or_else(|| {
                CoercionError::TypeMismatch(field_name.to_string(), "array".to_string())
            })?;
            let mut list = Vec::new();
            for item in arr {
                let coerced_item = coerce_field_value(&format!("{}[item]", field_name), &items.r#type, item)?;
                list.push(coerced_item);
            }
            Ok(Value::Array(list))
        }
    }
}

File: apps/api/src/callcraft_engine/coercion.py
from typing import Any, Dict, List
from .schema import FieldDefinition, PlatformDataType, ResponseSchema


class CoercionError(Exception):
    pass


def validate_and_coerce(schema: ResponseSchema, raw_value: Any, allow_missing_required: bool = False) -> Dict[str, Any]:
    if not isinstance(raw_value, dict):
        raise CoercionError("Root AI payload must be a JSON object")

    result_map: Dict[str, Any] = {}

    if "_ai_message" in raw_value:
        result_map["_ai_message"] = raw_value["_ai_message"]
    if "_executed_tools" in raw_value:
        result_map["_executed_tools"] = raw_value["_executed_tools"]

    for field_name, field_def in schema.properties.items():
        val = raw_value.get(field_name)

        if val is None:
            if field_def.required and not allow_missing_required:
                raise CoercionError(f"Missing required field: {field_name}")
            else:
                result_map[field_name] = (
                    field_def.default if field_def.default is not None else None
                )
        else:
            coerced_val = _coerce_field_value(field_name, field_def, val, allow_missing_required=allow_missing_required)
            result_map[field_name] = coerced_val

    return result_map


def _coerce_field_value(field_name: str, field_def: FieldDefinition, val: Any, allow_missing_required: bool = False) -> Any:
    dtype = field_def.type

    if dtype in (
        PlatformDataType.STRING,
        PlatformDataType.TEXT,
        PlatformDataType.EMAIL,
        PlatformDataType.PHONE,
        PlatformDataType.DATE,
        PlatformDataType.DATETIME,
        PlatformDataType.CURRENCY,
    ):
        if val is None:
            return None
        if isinstance(val, (str, int, float, bool)):
            return str(val).strip()
        raise CoercionError(f"Type mismatch for field {field_name}: expected string")

    elif dtype == PlatformDataType.INTEGER:
        if isinstance(val, int):
            return val
        if isinstance(val, float):
            return int(val)
        if isinstance(val, str):
            try:
                return int(val.strip())
            except ValueError:
                raise CoercionError(f"Type mismatch for field {field_name}: expected integer")
        raise CoercionError(f"Type mismatch for field {field_name}: expected integer")

    elif dtype == PlatformDataType.NUMBER:
        if isinstance(val, (int, float)):
            return float(val)
        if isinstance(val, str):
            try:
                return float(val.strip())
            except ValueError:
                raise CoercionError(f"Type mismatch for field {field_name}: expected number")
        raise CoercionError(f"Type mismatch for field {field_name}: expected number")

    elif dtype == PlatformDataType.BOOLEAN:
        if isinstance(val, bool):
            return val
        if isinstance(val, str):
            s = val.strip().lower()
            if s in ("true", "1", "yes"):
                return True
            if s in ("false", "0", "no"):
                return False
        raise CoercionError(f"Type mismatch for field {field_name}: expected boolean")

    elif dtype == PlatformDataType.ENUM:
        if not isinstance(val, str):
            raise CoercionError(f"Type mismatch for field {field_name}: expected enum string")
        str_val = val.strip()
        allowed = field_def.enum_values or []
        for a in allowed:
            if a.lower() == str_val.lower():
                return a
        allowed_str = ", ".join(allowed)
        raise CoercionError(
            f"Invalid enum value '{str_val}' for field {field_name}. Allowed values: {allowed_str}"
        )

    elif dtype == PlatformDataType.OBJECT:
        if not isinstance(val, dict):
            raise CoercionError(f"Type mismatch for field {field_name}: expected object")
        sub_props = field_def.properties or {}
        sub_schema = ResponseSchema(title=field_name, properties=sub_props)
        return validate_and_coerce(sub_schema, val, allow_missing_required=allow_missing_required)

    elif dtype == PlatformDataType.ARRAY:
        if not isinstance(val, list):
            raise CoercionError(f"Type mismatch for field {field_name}: expected array")
        res_list: List[Any] = []
        item_def = field_def.items or FieldDefinition(type=PlatformDataType.STRING)
        for i, item in enumerate(val):
            coerced_item = _coerce_field_value(f"{field_name}[{i}]", item_def, item, allow_missing_required=allow_missing_required)
            res_list.append(coerced_item)
        return res_list

    return val
